### PR TITLE
[IMP] sale_stock_margin: improve margin computation

### PR DESCRIPTION
**Pull request by std-odoo**
Improve standard price computation for FIFO product
===================================================
Before
------
If a product is FIFO, the standard price is the unit cost of the last released.

```
+ 1 product @ 10€
+ 1 product @ 15€
- 1 product
-> Actual standard price is 10€

```

After
-----
If a product is FIFO, the standard price of the ``product.product`` is the
unit cost of the next product which will  be released
```
+ 1 product @ 10€
+ 1 product @ 15€
- 1 product
-> Actual standard price is 15€

```

Improve the margin computation
==============================

Before
------
The margin stay unchanged after the validation of the delivery.

After
-----
The margin is updated when the delivery is validated.
To compute the margin for FIFO products, we use the value in the stock valuation layer as ``standard_price``.
```
e.g.:
    IN 1 product @ 10
    IN 2 product @ 20
    SO 4 products @ 40

    After delivery validation (after stock move creation), the margin is
    (40 * 4 - 10 - 20 - 20 - 20)
```

Task #2038418

### CODE REVIEW

File: addons/stock_account/models/product.py [hunk 1]
```diff
@@ -249,32 +249,36 @@ def _run_fifo(self, quantity, company):
...
             ('company_id', '=', company.id),
-        ])
+        ]))
         new_standard_price = 0
         tmp_value = 0  # to accumulate the value taken on the candidates
         for candidate in candidates:
             qty_taken_on_candidate = min(qty_to_take_on_candidates, candidate.remaining_qty)
 
             candidate_unit_cost = candidate.remaining_value / candidate.remaining_qty
-            new_standard_price = candidate_unit_cost
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: no need to move this line

File: addons/stock_account/models/product.py [hunk 2]
```diff
@@ -249,32 +249,36 @@ def _run_fifo(self, quantity, company):
...
             tmp_value += value_taken_on_candidate
+            new_standard_price = candidate_unit_cost
+
             if float_is_zero(qty_to_take_on_candidates, precision_rounding=self.uom_id.rounding):
+                if float_is_zero(candidate.remaining_qty, precision_rounding=self.uom_id.rounding):
+                    # continue the iteration on candidates thanks to iter()
+                    new_standard_price = next((
+                        candidate.remaining_value / candidate.remaining_qty
+                        for candidate in candidates if candidate.remaining_qty > 0
+                    ), new_standard_price)
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]:
``` diff

+                    # use `unit_cost` of the next candidate (if any) as `new_standard_price`
+                    if float_is_zero(candidate.remaining_qty, precision_rounding=self.uom_id.rounding):
+                        next_candidate = candidates.filtered(lambda svl: svl.remaining_qty > 0)
+                        new_standard_price = candidates and candidates[0].unit_cost or new_standard_price
```

File: addons/stock_account/models/stock_move.py [hunk 3]
```diff
@@ -34,7 +34,8 @@ def _get_price_unit(self):
         price_unit = self.price_unit
         # If the move is a return, use the original move's price unit.
         if self.origin_returned_move_id:
-            price_unit = self.origin_returned_move_id.price_unit
+            price_unit = self.origin_returned_move_id.stock_valuation_layer_ids.unit_cost
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_move.py]: using the initial price of the returned moved seems a good compromise to me
``` python
svls = self.origin_returned_move_id.stock_valuation_layer_ids
if svls:
    price_unit = svls[-1].unit_cost
```

File: addons/stock_account/tests/test_stockvaluation.py [hunk 4]
```diff
@@ -916,7 +916,7 @@ def test_fifo_perpetual_4(self):
 
         self.assertEqual(self.product1.standard_price, 16)
 
-        self.assertEqual(return_pick.move_lines.stock_valuation_layer_ids.unit_cost, 16)
+        self.assertEqual(round(return_pick.move_lines.stock_valuation_layer_ids.unit_cost, 1), 11.2)
```
Comment (sle-odoo) [inline addons/stock_account/tests/test_stockvaluation.py]: could you move this line and the previous line in the diff in their own commit? i think we'd need to backport them in v13 -_-

File: addons/stock_account/tests/test_stockvaluationlayer.py [hunk 5]
```diff
@@ -596,6 +595,27 @@ def test_dropship_1(self):
...
+        self._make_out_move(self.product1, 3, create_picking=True)
+        self.assertEqual(self.product1.standard_price, 23)
+
+    def test_standard_price_fifo_2(self):
+        self._make_in_move(self.product1, 5, unit_cost=17)
+        self._make_in_move(self.product1, 1, unit_cost=23)
+        self._make_out_move(self.product1, 4, create_picking=True)
+        self.assertEqual(self.product1.standard_price, 17)
+
+    def test_standard_price_fifo_3(self):
```
Comment (sle-odoo) [inline addons/stock_account/tests/test_stockvaluationlayer.py]: i would remove this test
users can't do this behaviour from the interface (You cannot validate a transfer if no quantites are reserved nor done. To force the transfer, switch in edit more and encode the done quantities.)

File: addons/stock_account/tests/test_stockvaluationlayer.py [hunk 6]
```diff
@@ -596,6 +595,27 @@ def test_dropship_1(self):
...
+        self._make_in_move(self.product1, 3, unit_cost=17)
+        self._make_in_move(self.product1, 1, unit_cost=23)
+        self._make_out_move(self.product1, 3, create_picking=True)
+        self.assertEqual(self.product1.standard_price, 23)
+
+    def test_standard_price_fifo_2(self):
+        self._make_in_move(self.product1, 5, unit_cost=17)
+        self._make_in_move(self.product1, 1, unit_cost=23)
+        self._make_out_move(self.product1, 4, create_picking=True)
+        self.assertEqual(self.product1.standard_price, 17)
```
Comment (sle-odoo) [inline addons/stock_account/tests/test_stockvaluationlayer.py]: can you move these two tests to test_stockvaluation.py?
this file is reserved to the implementation of our valuation spreadsheet

File: addons/sale_stock_margin/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+{
+    'name': "Sale Stock Margin",
+    'summary': '',
+    'description': 'Once the delivery is validated, update the cost on the SO to have an exact margin computation.',
+    'version': '0.1',
+    'depends': ['stock_account', 'sale_margin'],
```
Comment (sle-odoo) [inline addons/sale_stock_margin/__manifest__.py]: ['stock', 'sale_margin']
or ['sale_stock', 'sale_margin']

File: addons/sale_stock_margin/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    'installable': True,
+    'auto_install': True,
+    'data': [
+        'views/views.xml',
```
Comment (sle-odoo) [inline addons/sale_stock_margin/__manifest__.py]: to remove

File: addons/sale_stock_margin/models/sale_order_line.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.depends('product_id', 'company_id', 'currency_id', 'product_uom', 'move_ids', 'move_ids.stock_valuation_layer_ids', 'order_id.picking_ids.state')
+    def _compute_purchase_price(self):
```
Comment (sle-odoo) [inline addons/sale_stock_margin/models/sale_order_line.py]: you could update the original _compute_purchase_price
``` python
    @api.depends(lambda self: self._purchase_price_depends())
    def _compute_purchase_price(self):
        [...]
```

and only add the dependencies by inheriting _purchase_price_depends

File: addons/sale_stock_margin/models/sale_order_line.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        svl_linked_sol = self.filtered(lambda sol: sol.move_ids.stock_valuation_layer_ids and sol.qty_to_invoice)
+
+        for sale_order_line in svl_linked_sol:
+            sale_order_line.purchase_price = sale_order_line.product_id._compute_average_price(
+                sale_order_line.qty_invoiced,
+                sale_order_line.qty_to_invoice,
+                sale_order_line.move_ids)
+
+        super(SaleOrderLine, self-svl_linked_sol)._compute_purchase_price()
```
Comment (sle-odoo) [inline addons/sale_stock_margin/models/sale_order_line.py]:
``` python
lines_without_moves = self.browse()
for line in self:
    if not line.move_ids:
        lines_without_moves |= line
        continue
   line.purchase_price = line.product_id._compute_average_price(0, line.product_uom_qty, line.move_ids)
super(lines_without_moves)._compute_purchase_price()
```

File: addons/stock/models/stock_move.py [hunk 11]
```diff
@@ -276,7 +276,7 @@ def _compute_product_qty(self):
                 move.product_uom_qty, move.product_id.uom_id, rounding_method=rounding_method)
 
     def _get_move_lines(self):
-        """ This will return the move lines to consider when applying _quantity_done_compute on a stock.move. 
+        """ This will return the move lines to consider when applying _quantity_done_compute on a stock.move.
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: useless difffffff :)

File: addons/sale_stock_margin/tests/test_sale_stock_margin.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from datetime import timedelta
+
+from odoo.exceptions import UserError
+from odoo.fields import Datetime
+from odoo.tests.common import Form, TransactionCase, tagged
+
```
Comment (sle-odoo) [inline addons/sale_stock_margin/tests/test_sale_stock_margin.py]: can you reuse TestStockValuationCommon ?

File: addons/sale_margin/models/sale_order.py [hunk 13]
```diff
@@ -10,7 +10,10 @@ class SaleOrderLine(models.Model):
     margin = fields.Float(compute='_compute_margin', digits='Product Price', store=True)
     purchase_price = fields.Float(string='Cost', compute="_compute_purchase_price", digits='Product Price', store=True, readonly=False)
 
-    @api.depends('product_id', 'company_id', 'currency_id', 'product_uom')
+    def _purchase_price_depends(self):
+        return 'product_id', 'company_id', 'currency_id', 'product_uom'
```
Comment (sle-odoo) [inline addons/sale_margin/models/sale_order.py]: you could simply return a list

File: addons/sale_stock_margin/models/sale_order_line.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    def _purchase_price_depends(self):
+        return (*super()._purchase_price_depends(), 'move_ids', 'move_ids.stock_valuation_layer_ids', 'order_id.picking_ids.state')
```
Comment (sle-odoo) [inline addons/sale_stock_margin/models/sale_order_line.py]: return super(). _purchase_price_depends() + ['move_ids', ..]

File: addons/sale_stock_margin/tests/test_sale_stock_margin.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        sale_order.action_confirm()
+
+        self.assertEqual(order_line.purchase_price, 10)
+        self.assertEqual(round(sale_order.margin), 20)
+
+        sale_order.picking_ids.move_lines.quantity_done = 1
+        res = sale_order.picking_ids.button_validate()
+        Form(self.env[res['res_model']].with_context(res['context'])).save().process()
+
+        self.assertEqual(round(order_line.purchase_price, 1), 15)
```
Comment (sle-odoo) [inline addons/sale_stock_margin/tests/test_sale_stock_margin.py]: all your assertEqual(round
-> assertAlmostEqual

File: addons/stock_account/tests/test_stockvaluation.py [hunk 16]
```diff
@@ -103,6 +103,56 @@ def _get_stock_valuation_move_lines(self):
...
+        return out_move.with_context(svl=True)
+
+    def _get_new_fifo_product(self):
+        product = self.env['product.product'].create({
+            'name': 'product1',
+            'type': 'product',
+            'categ_id': self.env.ref('product.product_category_all').id,
+        })
+        product.product_tmpl_id.categ_id.property_cost_method = 'fifo'
+        return product
```
Comment (sle-odoo) [inline addons/stock_account/tests/test_stockvaluation.py]: a shame to introduce these helpers in this file and they are used in 3 of the 50 tests :( can you remove them et create the move the same way the others do? so this file is consistent

File: addons/stock_account/tests/test_stockvaluation.py [hunk 17]
```diff
@@ -3602,3 +3652,32 @@ def test_at_date_average_1(self):
...
+        self._make_in_move(product, 1, unit_cost=23)
+        self.assertEqual(product.standard_price, 17)
+        self._make_out_move(product, 4)
+        self.assertEqual(product.standard_price, 17)
+        self._make_out_move(product, 1)
+        self.assertEqual(product.standard_price, 23)
+        self._make_out_move(product, 1)
+        self.assertEqual(product.standard_price, 23)
+        self._make_in_move(product, 1, unit_cost=77)
+        self.assertEqual(product.standard_price, 77)
```
Comment (sle-odoo) [inline addons/stock_account/tests/test_stockvaluation.py]: can you rename these tests

test_fifo_standard_price_upate_1/2/3 and place them just after test_fifo_edit_done_move2 ?

**Comment by sle-odoo** [comment]
@Feyensv fyi

File: addons/purchase_stock/tests/test_fifo_price.py [hunk 18]
```diff
@@ -59,7 +59,7 @@ def test_00_test_fifo(self):
 
         # Check the standard price of the product (fifo icecream), that should have not changed
```
Comment (amoyaux) [inline addons/purchase_stock/tests/test_fifo_price.py]: remove the comment upper

File: addons/sale_stock_margin/models/sale_order_line.py [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    def _purchase_price_depends(self):
+        return super()._purchase_price_depends() + ['move_ids', 'move_ids.stock_valuation_layer_ids', 'order_id.picking_ids.state']
```
Comment (sle-odoo) [inline addons/sale_stock_margin/models/sale_order_line.py]: move_ids, move_ids.state should be enough

File: addons/stock_account/models/stock_move.py [hunk 20]
```diff
@@ -328,6 +328,13 @@ def product_price_update_before_done(self, forced_qty=None):
             move.product_id.with_company(move.company_id.id).with_context(disable_auto_svl=True).sudo().write({'standard_price': new_std_price})
             std_price_update[move.company_id.id, move.product_id.id] = new_std_price
 
+        # adapt standard price on incomming moves if the product cost_method is 'fifo'
+        for move in self.filtered(lambda move:
+                                  move._is_in()
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_move.py]: _is_in is not needed here, self already contains in move

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r-
simplify depends

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**